src/Compilers/Shared/BuildServerConnection.cs

```diff
@@ -595,4 +621,205 @@ public static string GetTempPath(string workingDir)
+            if (String.IsNullOrWhiteSpace(homeDirectory))
+                homeDirectory = Path.GetTempPath();
+            var result = Path.Combine(homeDirectory, ".roslyn");
+            if (!Directory.Exists(result))
+                Directory.CreateDirectory(result);
+            return result;
+        }
+
+        public FileMutex (string name)
+        {
+            FilePath = Path.Combine(GetMutexDirectory(), name);
+            try
+            {
+                Stream = new FileStream(FilePath, FileMode.Open, FileAccess.ReadWrite, FileShare.None);
+                OwnsMutex = false;
+            }
+            catch (FileNotFoundException)
+            {
+                Stream = new FileStream(FilePath, FileMode.CreateNew, FileAccess.ReadWrite, FileShare.None);
+                OwnsMutex = true;
+            }
+        }
+
+        public bool TryLock (int timeoutMs)
+        {
+            if (IsLocked)
+                throw new InvalidOperationException("Lock already held");
+
+            var sw = Stopwatch.StartNew();
+            do
+            {
+                try
+                {
+                    Stream.Lock(0, 0);
+                    IsLocked = true;
+                    return true;
+                }
+                catch (IOException)
+                {
+                    // Lock currently held by someone else
+                    Thread.Sleep(0);
+                }
+                catch (Exception)
+                {
+                    // Something else went wrong.
+                    return false;
+                }
+            } while (sw.ElapsedMilliseconds < timeoutMs);
+
+            return false;
+        }
+
+        public void Unlock ()
+        {
+            if (!IsLocked)
+                return;
+            Stream.Unlock(0, 0);
+            IsLocked = false;
+        }
+
+        public void Dispose ()
+        {
+            var wasLocked = IsLocked;
+            if (wasLocked)
+                Unlock();
+            Stream.Dispose();
+            // We do not delete the lock file here because there is no reliable way to perform a
+            //  'delete if no one has the file open' operation atomically on *nix. This is a leak.
+        }
+    }
+
+    internal sealed class ServerNamedMutex : IServerMutex
+    {
+        private readonly Mutex Mutex;
+
+        public bool IsDisposed { get; private set; }
+        public bool IsLocked { get; private set; }
+
+        public ServerNamedMutex (string mutexName, bool createNewLocked, out bool createdNew)
+        {
+            if (createNewLocked) {
+                Mutex = new Mutex(
+                    initiallyOwned: true,
+                    name: mutexName,
+                    createdNew: out createdNew
+                );
+                if (createdNew)
+                    IsLocked = true;
+            } else {
+                Mutex.TryOpenExisting(mutexName, out Mutex);
```

Comment:
I don't think we should be conflating the two different ways of opening the `Mutex` here. Rather I think we should just be forwarding this to the `Mutex` constructor as we did before. The case for `TryOpenExisting` is for checking if the `Mutex` existed. I think we should roughly use that code path as is and keep it separate from this one. Possibly just add `ServerNamedMutex.WasOpen(string mutexName)` and put the original code in there.